catalog/CatalogRelation.hpp

```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
   }
 
   /**
-   * @brief Check if an index by the given name exists already
-   *
-   * TODO (ssaurabh): Provide a better interface to query indices.
+   * @brief Check whether an index with the given exists or not.
    *
-   * @param The query index name
-   * @return true, if the index with the given name exists, otherwise false.
+   * @param index_name Name of the index to be checked.
+   * @return Whether the index exists or not.
    **/
-  bool hasIndexWithName(const std::string& name) {
-    SpinSharedMutexExclusiveLock<false> lock(indices_mutex);
-    std::vector<std::string>::iterator it = std::find(indices_.begin(), indices_.end(), name);
-    if (it != indices_.end()) {
-      return true;
-    } else {
+  bool hasIndexWithName(const std::string &index_name) const {
+    if (index_scheme_.get() == nullptr) {
       return false;
+    } else {
+      return index_scheme_->hasIndexWithName(index_name);
     }
   }
 
   /**
-   * @brief Create an index with a given name
+   * @brief Check whether an index with the given description
+   *        containing the same attribute id and index type
+   *        exists or not in the index map.
    *
-   * TODO (ssaurabh): This is only a placeholder implementation
-   * that just adds an index name into an indicies vector. This
-   * will be replaced by a full-fledged index data structure.
+   * @param index_descripton Index Description to check against.
+   * @return Whether a similar index description was already defined or not.
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    if (index_scheme_.get() == nullptr) {
+      return false;
+    } else {
+      return index_scheme_->hasIndexWithDescription(index_description);
+    }
+  }
+
+  /**
+   * @brief Add an index to the index_map.
    *
-   * @param The index to create
+   * @param index_name Name of the index to be added.
+   * @param index_description Corresponding description of the index.
+   * @return Whether the index was added successfully or not
    **/
-  void addIndex(const std::string& name) {
-    SpinSharedMutexExclusiveLock<false> lock(indices_mutex);
-    indices_.emplace_back(name);
+  bool addIndex(const std::string &index_name, const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    // create an index_scheme, if it does not exist
+    if (index_scheme_.get() == nullptr) {
+      index_scheme_.reset(new IndexScheme());
+    }
+
+    // verify that index with the given name does not exist
+    if (index_scheme_->hasIndexWithName(index_name)) {
+      return false;
+    }
+
+    // verify that index with similar description does not exist
+    std::vector<IndexSubBlockDescription>::const_iterator itr;
```

Comment:
Remove this line, and use `auto` in the for-loop.

Btw, we typically use `it` when shortening `iterator`, instead of `itr`, while `cit` for `const_iterator`.